Add incident report for deploy issue on 2021-06-29 (#290)
usdigitalresponse/univaf: docs/incidents/2021-06-29--actions-budget-terraform-state.md

Both of these issues from poor coordination and lack of a clear process around infrastructure management, and definitely need more work to come up with long-term solutions. (See [action items](#action-items) at the end of this report.)

## Timeline

*(All times US/Pacific, 24-hour clock.)*

### 2021-06-29 18:30

@Mr0grog completes hotfix [#275][issue-275], but discovers actions are not running and all our GitHub checks are failing. He starts trying to figure out what is going wrong and whether this is specific to our repo or a GitHub-wide problem.

### 2021-06-29 18:36

Posted a request for help in #team-ops-and-tools in case someone is with better access to the GitHub org account can figure out the issue more clearly. (@Mr0grog does not anticipate this will be very fruitful since it’s after hours, and there’s no clear ops on-call/escalation plan as far as he knows.)

### 2021-06-29 19:04

Realized this is a billing issue. Error messages about payment failures are shown on below the fold on the workflow summary screen (not on individual job logs). @Mr0grog updates thread in #team-ops-and-tools with info and a more specific ask about payment.

### 2021-06-29 20:41

> @Mr0grog gave up on resolving the billing issue and pinged @jaronoff97 for help reviewing the hotfix since automated checks aren’t running.

Reviewer: ```suggestion
@Mr0grog gives up on resolving the billing issue and pings @jaronoff97 for help reviewing the hotfix since automated checks aren’t running.
```

### 2021-06-29 20:43

@jaronoff97 reviews the PR and @Mr0grog merges and starts building & uploading Docker images manually.

### 2021-06-29 21:14

> @Mr0grog rebuilds and uploads on remote infrastructure. His home internet connection is too slow to upload the build Docker images with any reasonable speed.

Reviewer: Worth adding a little detail here about the remote infrastructure you used?

Author: How’s 26ecfe6def0e61ba3a42365a88edbe1419c0eb84?

### 2021-06-29 21:36

> Tried to deploy new images via Terraform, but the Terraform plan shows that the bastion security group is going to be destroyed. @Mr0grog is no longer sure this deploy is safe, asks on Slack if @jaronoff97 knows what’s up, and starts digging into it.

Reviewer: ```suggestion
@Mr0grog tries to deploy new images via Terraform, but the Terraform plan shows that the bastion security group is going to be destroyed. @Mr0grog is no longer sure this deploy is safe, asks on Slack if @jaronoff97 knows what’s up, and starts digging into it.
```

### 2021-06-29 22:32

After digging around online, @Mr0grog discovers how to manually change state via the Terraform API in this article: https://learn.hashicorp.com/tutorials/terraform/cloud-state-api

He downloads the state from the last deploy, modifies its sequence number, signs it, and uploads it as the new state via the API, then triggers a new plan from the `main` git branch in Terraform Cloud.

### 2021-06-29 22:34

The plan finishes and shows sensible output that isn’t destroying resources. @Mr0grog applies.

### 2021-06-29 22:40

Infrastructure in AWS is updated and hotfix is finally deployed.

The original bug @Mr0grog was trying to fix is resolved, but automatic builds are still broken. Terraform is back in a good state, but clarification on the exact cause and whether further cleanup is necessary is still needed from @jaronoff97 in the morning. @Mr0grog posts a quick summary and questions in Slack and goes to bed.

---

Post-resolution:

### 2021-06-30 06:17

@jaronoff confirms diagnosis and fix were correct, kicks off discussion about processes to avoid this in the future.

### 2021-06-30 06:17

> Alex suggests posting about billing issues in #questions-tools instead of #team-ops-and-tools.

Reviewer: ```suggestion
@alexallain suggests posting about billing issues in #questions-tools instead of #team-ops-and-tools.
```

### 2021-06-30 10:33

> Paul adds some credit to USDR GitHub account to get actions running again. Automated tests, builds, and Terraform deploys are all now functional again.

Reviewer: ```suggestion
@paulschreiber adds some credit to USDR GitHub account to get actions running again. Automated tests, builds, and Terraform deploys are all now functional again.
```

### 2021-06-29 18:36

@Mr0grog posts a request for help in #team-ops-and-tools in case someone is with better access to the GitHub org account can figure out the issue more clearly. (@Mr0grog does not anticipate this will be very fruitful since it’s after hours, and there’s no clear ops on-call/escalation plan as far as he knows.)

### 2021-06-29 19:04

@Mr0grog realizes this is a billing issue. Error messages about payment failures are shown on below the fold on the workflow summary screen (not on individual job logs). @Mr0grog updates thread in #team-ops-and-tools with info and a more specific ask about payment.

### 2021-06-29 20:41

@Mr0grog gives up on resolving the billing issue and pings @jaronoff97 for help reviewing the hotfix since automated checks aren’t running.

### 2021-06-29 20:43

@jaronoff97 reviews the PR and @Mr0grog merges and starts building & uploading Docker images manually on his local machine.

### 2021-06-29 21:14

@Mr0grog gives up on uploading locally build images (his home internet connection is too slow to upload with any reasonable speed). He rebuilds and uploads from remote infrastructure instead:

> - He first begins building on an EC2 machine in his person AWS account.

Reviewer: ```suggestion
- He first begins building on an EC2 machine in his personal AWS account.
```
